### DISCUSSION TITLE
Unit testing on PlatformIO

### DISCUSSION THREAD

**Discussion by OOHehir** [LIFsCode/ELOC-3.0#21] [General]
Hello,

I was attempting to implement unit testing on a desktop (native) build for the TensorFlow stuff but its a bit of a struggle, lots of build errors as everything in the lib folder gets built. Its seems to be a known [issue](https://community.platformio.org/t/problem-with-native-unit-testing-on-a-large-r-project/26683/7), I was wondering if anyone has a solution?

Thanks,

Owen

**Comment by LIFsCode** [comment]
@OOHehir yes I'm also lacking some unit testing in the ELOC project. 
There is support for the [qemu emulator](https://docs.platformio.org/en/latest/advanced/unit-testing/simulators/qemu.html) in platformio. But I have not used it so far. No idea how easy it is to setup and how the performance is

**Comment by OOHehir** [comment]
@LIFsCode Unfortunately it looks like there is only very [limited support](https://docs.platformio.org/en/latest/plus/debug-tools/qemu.html#frameworks) for this emulator. 

I came across [this](https://vcon.io/automated-firmware-tests/) previously, could be a useful method of running continuous automated tests on the actual hardware.

**Comment by OOHehir** [comment]
@LIFsCode @EDsteve Just an update on this, I managed to get the unit testing environment up & running. Its documented [here](https://github.com/LIFsCode/ELOC-3.0/blob/master/eloc610LowPowerPartition/test/README.md).  As it states it supports generic/ desktop testing & target testing.

I had to reorganize the platform.ini file to support the testing. Hopefully won't cause any issues.

There's a separate topic of autonomous programming & reset of the ELOC board that I'll open another discussion for.

**Comment by OOHehir** [comment]
I'll close this as feature added &  #88 enables testing.
